library/pkparse.c

```diff
@@ -1320,21 +1313,23 @@ static int pk_parse_key_sec1_der(mbedtls_pk_context *pk,
 
     if (!pubkey_done) {
         if ((ret = pk_derive_public_key(pk, d, d_len, f_rng, p_rng)) != 0) {
-#if !defined(MBEDTLS_PK_USE_PSA_EC_DATA)
-            mbedtls_ecp_keypair_free(eck);
-#endif
-            return ret;
+            goto exit;
         }
     }
 
 #if !defined(MBEDTLS_PK_USE_PSA_EC_DATA)
     if ((ret = mbedtls_ecp_check_privkey(&eck->grp, &eck->d)) != 0) {
-        mbedtls_ecp_keypair_free(eck);
-        return ret;
+        goto exit;
     }
 #endif /* !MBEDTLS_PK_USE_PSA_EC_DATA */
 
-    return 0;
+exit:
+#if !defined(MBEDTLS_PK_USE_PSA_EC_DATA)
```

Comment:
I was going to question why we want to skip this when `PK_USE_PSA_EC_DATA`. But I also checked all calls to `pk_parse_key_sec1_der()` just to make sure, and it turns out all 3 of them already call `mbedtls_pk_free()` on failure. So,  actually we don't need to clean here.

So now I'm changing my mind: we can keep the `return`s rather that `goto exit` and just stop calling `ecp_keypair_free()` or `pk_free()` on failure - our callers are already doing taking care of it (and since the callers are also doing the `pk_setup()`, I think it makes more sense to keep the matching `pk_free()` in the callers).

Reply:
Note: also, this is a `static` function, so it's OK to rely on callers doing the right thing. If this were a public function, we'd probably want to keep the `pk_free()` inside the function as a precaution against misuse. (Just to clarify in case a similar question arises with a public function in the future.)

Reply:
> I changed my mind about a suggestion I made, sorry 

No problem at all! Actually thanks for your analysis, it's something I didn't thought about, but it seems absolutely fine to me. I'll fix this

Reply:
While doing this change I realized that there were also other `mbedtls_ecp_keypair_free()` in `pkparse.c` that could be removed for the same reason so I did it all at once in the new commit https://github.com/Mbed-TLS/mbedtls/pull/7682/commits/40df03f2c9523fb2e93cbbdc2e798e51e1d083ab